### PR TITLE
Add Test Client and Framework for e2e tests

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
This PR add a `TestClient` and Framework. The `TestClient` is a wrapped kubevirt client that exposes all client functionality and can be used as a normal client. In addition it provides a mechanism that track the non-namespaces resources created or updated during e2e tests, and functions to rollback (cleanup) them at the end of the execution.
This` Framework` supports common operations used by e2e tests; it will keep a client & clean the cluster for you. It is not necessary anymore to clean the cluster on `BeforeEach` of all tests if you use the framework because it cleans it for you. It contains a call to `tests.BeforeTestCleanup()` on its `BeforeEach()`. 
It also initializes the `TestClient` that can be used everywhere: in this way, writing tests, you have not to retrieve the `TestClient` because it creates one for you, you can simply use it. 
Looking at the tests files there where different cases where`tests.BeforeTestCleanup()` was defined inside `BeforeEach()` of the outer `Describe` block and also in the inner ones(`Context, It, Describe`..). Since `BeforeEach()` is executed for each specs it causes a multiple calls to `tests.BeforeTestCleanup()` function.
In other cases, it was not defined in any of the `BeforeEach()` causing a possible dependency between tests.
The union of these two gives the following advantages:
1. Make sure that each test will run in a cleaned cluster
2. Provides to you a `TestClient` that can be used without the necessity of retrieving ones
3. Gives the possibility to write a `BeforeEach()` that is specific for that set of tests

Since it cleans the cluster before each test this ensures that every test runs in the same cluster conditions.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Only a subset of the tests is affected by changes in order to monitoring issues.
The tests were chosen on the basis of whether or not they contained changes to non-namespaces resources tracked by the test client:

- perstistenVolume CREATE
- node CREATE, UPDATE, PATCH
- namespace CREATE
- migrationPolicy CREATE

**Release note**:

```release-note
Add TestClient to easily rollback non-namespaced resources created or updated during e2e tests
Add Framework to support common operations used by functional tests
```

### CODE REVIEW

**Comment by fossedihelm** [comment]
> I just had quick look and I think it would help me if we split the changes to tests from the first commit. It also is logical imho. WDYT?

@xpivarc Thanks, you are definitely right. Will do

File: tests/framework/client/persistentvolume.go
Comment (fossedihelm) [reply tests/framework/client/persistentvolume.go:17]: The idea was to keep each "tracked" information in its own context, but if you think that keep it all in the client will increase the readability, or for some other reasons, I have no problem to change it :)

**Comment by fossedihelm** [comment]
@xpivarc I arrange the commits in the following order:
- Framework creation
- TestClient creation
- Changes to e2e tests

File: tests/framework/client/node.go
Comment (xpivarc) [inline tests/framework/client/node.go:53]: We could also check if we need to update the node.

File: tests/framework/client/node.go
Comment (xpivarc) [inline tests/framework/client/node.go:123]: I would probably negate this. If we are changing created node for a test we don't need to keep the previous state.

File: tests/framework/client/node.go
Comment (xpivarc) [inline tests/framework/client/node.go:130]: Do we modify the spec of nodes currently?

File: tests/framework/client/persistentvolume.go
Comment (xpivarc) [reply tests/framework/client/persistentvolume.go:17]: I think this is fine as far we don't use client concurently, right?

File: tests/framework/framework/framework.go
Comment (fossedihelm) [reply tests/framework/framework/framework.go:41]: No, you didn't miss anything. Currently it is never used. I think that it could be useful for future improvements (i.g. create test specific labels). I can delete it if you think that is better to "add it when we need it"

File: tests/framework/framework/framework.go
Comment (fossedihelm) [reply tests/framework/framework/framework.go:85]: Sorry I don't understand. This function call makes the cluster clean before all tests, instead of declare it in each single test file. Why do we have to remove it?

File: tests/framework/client/node.go
Comment (fossedihelm) [reply tests/framework/client/node.go:123]: Probably the comment is misleading. Here we check that the node is gonna to be modified is not a previous created one:
if we want to patch a node `x` and `x` was manually created during the test I'm not going to save the state of `x` because I know that `x` will be deleted at the end of the test.

File: tests/framework/client/node.go
Comment (fossedihelm) [reply tests/framework/client/node.go:130]: In the `Spec` we have `Unschedulable` and `Taints` fields, and in some tests we modify them. Instead of store the specific fields I opted to store all the `Spec`

File: tests/framework/client/node.go
Comment (xpivarc) [reply tests/framework/client/node.go:123]: Exactly I was just suggesting doing `if _, exist := createdNodes[name]; exist { //node will be deleted return}`. Just suggestion.

File: tests/framework/client/persistentvolume.go
Comment (fossedihelm) [reply tests/framework/client/persistentvolume.go:17]: The idea is not to use another `client` concurrently because otherwise the `testClient` is bypassed. This has to be forethought of the developers and the reviewers. On the other hand, also if you use the `kubectl` or the `RestClient` you can bypass it.

File: tests/framework/client/client.go
Comment (fossedihelm) [reply tests/framework/client/client.go:30]: In the client we initialize the `coreV1Client` during [https://github.com/kubevirt/kubevirt/blob/a6a84d0388b12d8c812e19cb7cc88f5193e42bdf/staging/src/kubevirt.io/client-go/kubecli/kubecli.go#L146](https://github.com/kubevirt/kubevirt/blob/a6a84d0388b12d8c812e19cb7cc88f5193e42bdf/staging/src/kubevirt.io/client-go/kubecli/kubecli.go#L146) and it is done only once (because `CoreV1()` is managed by k8s itself)
I don't think it is necessary do it in the client itself, because it is related to the fact that I have to override the `CoreV1()` function. I don't know if this reply to your question.

File: tests/framework/client/node.go
Comment (fossedihelm) [reply tests/framework/client/node.go:123]: Oh I got it. Sorry, my bad. Agree

**Comment by fossedihelm** [comment]
@xpivarc Thanks for your review. I reply to your comments. Thanks

File: tests/framework/client/node.go
Comment (fossedihelm) [reply tests/framework/client/node.go]: Done

File: tests/framework/client/node.go
Comment (fossedihelm) [reply tests/framework/client/node.go:55]: I add a check on clean as you suggested, thanks

File: tests/framework/client/node.go
Comment (fossedihelm) [reply tests/framework/client/node.go]: I modified the `nodeSpec` struct, using `Unschedulable` and `Taints` directly, avoiding storing other stuff we are not interested in. Thanks

**Comment by fossedihelm** [comment]
@xpivarc Pushed requested changes and edited some `tests/utils.go` functions: passing `virtClient` as parameter we ensure to not bypass non-namespaces resources creation. Thanks

**Comment by fossedihelm** [comment]
I edited last commit in order to start with a subset of the tests to iterate on possible issues.
The tests were chosen on the basis of whether or not they contained changes to non-namespaces resources tracked by the test client:

- perstistenVolume CREATE
- node CREATE, UPDATE, PATCH
- namespace CREATE
- migrationPolicy CREATE

File: tests/framework/client/client.go
Comment (iholder-redhat) [inline tests/framework/client/client.go:20]: nit: would change to:
```go
// TestClient is a wrapped kubevirt client that exposes all client functionality
// and can be used as a normal client. It provides a mechanism that track
// non-namespaced resources created or updated during e2e tests and functions to rollback them at the end of the execution.
type TestClient struct {
```

that is:
- "TestClient" is now part of the explanation
- added space between slashes and text
- removed empty line between doc and struct declarazation

File: tests/framework/client/client.go
Comment (iholder-redhat) [inline tests/framework/client/client.go:52]: nit: leftover

File: tests/framework/client/migrationpolicy.go
Comment (iholder-redhat) [inline tests/framework/client/migrationpolicy.go:27]: To avoid having to re-implement all of the interface's, you better [embed](https://travix.io/type-embedding-in-go-ba40dd4264df) the migration policy client.

In other words, if you'd define the struct that way:
```go
type migrationPolicies struct {
	migrationsv1.MigrationPolicyInterface
	client          rest.Interface
}

// newMigrationPolicies returns a MigrationPolicies
func newMigrationPolicies(c *TestMigrationPolicyClient) *migrationPolicies {
	return &migrationPolicies{
		MigrationPolicyInterface: c.client.MigrationPolicies(),
		client:                   c.client.RESTClient(),
	}
}
```

Then you would be able to simply:
```go
var c *migrationPolicies = ...
c.Get()
```

And you wouldn't need to recreate all Get, List, Create, etc methods.

This is also easier to implement but would make sure we're up to date with client's methods :)

Comment (iholder-redhat) [reply tests/framework/client/migrationpolicy.go:27]: p.s. you can still override whatever methods you need, like Create()
and obviously the above comment applies to every client your created, not just in this file

File: tests/framework/client/migrationpolicy.go
Comment (iholder-redhat) [inline tests/framework/client/migrationpolicy.go:75]: Maybe in delete we would like to take the policy out of `createdMigrationPolicies`?
But on the other hand this is just a pointer and shouldn't really effect test performance / usa a lot of memory

File: tests/framework/client/client.go
Comment (iholder-redhat) [inline tests/framework/client/client.go:63]: What I'd suggest here is to have a `[]CleanableResource` slice defined with the vars above.
Then, when each client is created (e.g. when calling `MigrationPolicy()`) you can use the `sync.One` to append the client to this slice.

Finally when you get here, you can simply iterate the slice and call each element's "Clean()" method.

File: tests/framework/framework/framework.go
Comment (iholder-redhat) [inline tests/framework/framework/framework.go:112]: My IDE says:
```
Deprecated: CurrentGinkgoTestDescription has been replaced with CurrentSpecReport.
Use CurrentSpecReport() instead. You can learn more here: https://onsi.github.io/ginkgo/#getting-a-report-for-the-current-spec
```

**Review by iholder-redhat — COMMENTED** [review]
Thanks @fossedihelm!!
This is a great idea! Well done! Happy to see it.

p.s. there are a lot more `tests.BeforeTestCleanup()` calls left, but I guess we can remove the old usage step by step.

File: tests/framework/client/client.go
Comment (fossedihelm) [reply tests/framework/client/client.go:20]: Great, thanks :+1:

File: tests/framework/client/migrationpolicy.go
Comment (fossedihelm) [reply tests/framework/client/migrationpolicy.go:27]: You're totally right. I had some troubles trying to do that, but now it works. Thank you so much

File: tests/framework/client/migrationpolicy.go
Comment (fossedihelm) [reply tests/framework/client/migrationpolicy.go:75]: We could, but since  `Clean()` ignore notFound errors, I think we can leave it as is. At most we try to remove it twice

File: tests/framework/client/client.go
Comment (fossedihelm) [reply tests/framework/client/client.go:63]: Thanks, I like it

File: tests/framework/client/migrationpolicy.go
Comment (iholder-redhat) [reply tests/framework/client/migrationpolicy.go:75]: Right, but it still has two disadvantages:
* Memory leak (although very small since we're only dealing with pointers)
* Overhead (since you make an API call for nothing)

Of course these two are far from being significant. Your call :)

Comment (fossedihelm) [reply tests/framework/client/migrationpolicy.go:75]: Agreed, In this case I think I will move `createdMigrationPolicies` from slice to map in order to not iterate over it for every `Delete`. WDYT?

Comment (iholder-redhat) [reply tests/framework/client/migrationpolicy.go:75]: Sounds good to me :+1: 
I would also remove it from the map if I know a policy is already deleted

File: tests/framework/framework/framework.go
Comment (fossedihelm) [reply tests/framework/framework/framework.go]: Fixed, It was inherited from upgrading to ginkgo2

**Comment by iholder-redhat** [comment]
Thanks a lot @fossedihelm! Great to see this! Awesome work.
/lgtm

**Comment by fossedihelm** [comment]
> Thanks @fossedihelm!! This is a great idea! Well done! Happy to see it.
> 
> p.s. there are a lot more `tests.BeforeTestCleanup()` calls left, but I guess we can remove the old usage step by step.

@iholder-redhat Yes, the approach was to starting with a subset of tests and, once we are sure enough, to extend it to all other tests.
I pushed the requested changes, that can be summarized as follow:

- Changed `TestClient` description
- Used `embeded` interfaces, which allows to override the only necessary functions
- Add handler for `Delete` api call
- Changed `Clean()` behavior using `[]CleanableResources`
- Update deprecated function call in `framework.go` file

**Comment by fossedihelm** [comment]
@iholder-redhat 

- Switched from `Update` to `Patch` in `node.go` https://github.com/kubevirt/kubevirt/blob/cdc7130e8a98cd5384c183cde3636ac3ee74b78b/tests/framework/client/node.go#L76

- Using `Expect()` instead of `panicOnError`

**Comment by fossedihelm** [comment]
Rebased

**Comment by iholder-redhat** [comment]
> Rebased
re-lgtm
/lgtm

**Review by jean-edouard — COMMENTED** [review]
/approve

**Comment by EdDev** [comment]
This seems like a proposal to change the way we write e2e tests in general.
If this is indeed the case, IMO we need feedback from a wide audience and an improve the exposure of this to the other contributors.

Unlike previous changes to the e2e tests, the PR description hints it is a major logical change.
Could you please present it in some forum and ask for feedback from more maintainers and contributors?

Placing a hold to have a chance to review this and understand the implications. I hope more will join in to review this.
/hold

**Review by EdDev — COMMENTED** [review]
I like the general approach of introducing a "framework" to help with our complex e2e tests.
I saw this used in k8s e2e tests and it seemed to have made their tests much more compact and clean.

I would start with a request to introduce this through a design proposal and not directly as a PR.
It affects how tests should be written, and how existing tests could be made cleaner. All this requires clear motivation, goals, structure and phases to be properly introduced.
Other SIG/s need to contribute their own helpers and make sure they can be used, so high level review at the planning stage can be very useful.

To the change proposal itself: I managed to go over only the first commit.
I found it challenging to review, mainly because it introduced many of the client functionality/abstraction without seeing it used. I had no choice but to give up and look at the PR as one change, which made it even harder.
In this regard, I think there needs to be a strategy where the changes are introduced gradually, one functionality at a time, in small portions.

Although I don't feel I even stated to review it from top to bottom, I do see one issue that in k8s community it has been raised several times:
- While a "framework" provides a skeleton and structure to how a test should be written (e.g. abstracting cleanup), the functional helpers (e.g. the kubevirt-client) are a separate part which can be considered as libraries. In practice, what I suggest here is to take out the test-client out of the "framework" and let users use it directly even without the framework.
This will also serve well in introducing it in small commits (and possibly a smaller PR).

Regarding the framework, I am still not deep into the details as I need to fetch locally the PR and review directly. It is too big to be reviewed on GitHub diff (e.g. I tried to look for usages of some methods, but it did not find any, probably because some files are not shown in the diff).
But perhaps I can ask this even before I start: Have you used in this framework GinkGo v2 capabilities (or considered them)?

**Review by EdDev — CHANGES_REQUESTED** [review]
I started looking at the solution locally and I must admit it is challenging to comprehend.

Would you consider simplifying this with explicit creation methods?
i.e. `f.CreateNamespace(...)`.

With the existing tooling from gingko, the teardown can be as simple as:
```
ns := f.CreateNamespace(...)
DeferCleanup(f.DeleteNamespace, ns)
```

I would strive to simplify everything and make it explicit when used in the tests, not hide it under multiple interfaces.

**Comment by fossedihelm** [comment]
@EdDev Thanks for your review and the time spent on this! Few replies inline.

> I like the general approach of introducing a "framework" to help with our complex e2e tests. I saw this used in k8s e2e tests and it seemed to have made their tests much more compact and clean.
> 
> I would start with a request to introduce this through a design proposal and not directly as a PR. It affects how tests should be written, and how existing tests could be made cleaner. All this requires clear motivation, goals, structure and phases to be properly introduced. Other SIG/s need to contribute their own helpers and make sure they can be used, so high level review at the planning stage can be very useful.

Agreed, I will post a design proposal

> To the change proposal itself: I managed to go over only the first commit. I found it challenging to review, mainly because it introduced many of the client functionality/abstraction without seeing it used. I had no choice but to give up and look at the PR as one change, which made it even harder. In this regard, I think there needs to be a strategy where the changes are introduced gradually, one functionality at a time, in small portions.

Can you explain this part? Talking about portions, are you speaking about testClient vs Framework? or smaller portions of testClient, i.e. node/namespace/migrationPolicy/persistentVolume? In the first case I agree with you, in the second one, I think that the test client watching only one resource could results in a not completed functionality creating confusion about which resource is tracked and which is not.
 
> Although I don't feel I even stated to review it from top to bottom, I do see one issue that in k8s community it has been raised several times:
> 
> * While a "framework" provides a skeleton and structure to how a test should be written (e.g. abstracting cleanup), the functional helpers (e.g. the kubevirt-client) are a separate part which can be considered as libraries. In practice, what I suggest here is to take out the test-client out of the "framework" and let users use it directly even without the framework.
>   This will also serve well in introducing it in small commits (and possibly a smaller PR).

I will adjust this PR in the introduction on the only testClient (maybe I will create a new one linking this one), and a design proposal for the framework. WDYT?
 
> Regarding the framework, I am still not deep into the details as I need to fetch locally the PR and review directly. It is too big to be reviewed on GitHub diff (e.g. I tried to look for usages of some methods, but it did not find any, probably because some files are not shown in the diff). But perhaps I can ask this even before I start: Have you used in this framework GinkGo v2 capabilities (or considered them)?

Nope, this PR started before the upgrade to GinkGo v2 and then there was only rebase. I will appreciate any suggestion about it, thanks!

I will address your comment in the new PR, is it ok for you? Thanks

File: tests/framework/client/client.go
Comment (fossedihelm) [reply tests/framework/client/client.go:1]: You're right thanks!

File: tests/framework/framework/framework.go
Comment (brybacki) [reply tests/framework/framework/framework.go:85]: Is there a way to design it so failed test reporting (with log dump) runs after tests, and cleanup is done later, after that? I suppose this should work, because we use `JustAfterEach` for reporting and we can use `AfterEach` for cleanup. 

Is there a problem that forces us to clean before test? Is it better to clean before?

Comment (brybacki) [reply tests/framework/framework/framework.go:85]: I have started reading a discussion on kubevirt-dev. I can see it mentioned as "an intermediate solution". So maybe I need to finish reading the discussion before reviewing this PR.

Comment (fossedihelm) [reply tests/framework/framework/framework.go:85]: @brybacki As you said we collect reporting data in the `JustAfterEach` so yes, basically we can cleanup in the `AfterEach`.
Right, the `Framework` was intended, at least by me, as an intermediate solution that should be extended step by step introducing functionalities. But it's ok for me to drop it from this PR and discuss which functionalities we want in the proposal design.

**Comment by fossedihelm** [comment]
> I started looking at the solution locally and I must admit it is challenging to comprehend.
> 
> Would you consider simplifying this with explicit creation methods? i.e. `f.CreateNamespace(...)`.
> 
> With the existing tooling from gingko, the teardown can be as simple as:
> 
> ```
> ns := f.CreateNamespace(...)
> DeferCleanup(f.DeleteNamespace, ns)
> ```
> 
> I would strive to simplify everything and make it explicit when used in the tests, not hide it under multiple interfaces.

@EdDev I'm not clear what you're referring to: since you are mentioning interfaces I think you are referring to the testClient,
but your example shows usage of the Framework.
The test client was intended on purpose to automatically track changes to the cluster that could generate leftovers.
Here you are suggesting manual handling through `DeferCleanup`.
The two are probably incompatible and so I am beginning to wonder if the testClient in this respect has any use. Or did I misunderstand? Thanks

File: tests/framework/client/node.go
Comment (maiqueb) [inline tests/framework/client/node.go:16]: this import path should be changed imo - v12 says nothing informative about the pkg.

**Pull request closed without merging**
